File: src/lib/agente-tools.ts
/**
 * src/lib/agente-tools.ts
 * ─────────────────────────────────────────────────────────
 * Definición de tools del agente + ejecutor.
 *
 * Las tools llaman a los endpoints Next.js existentes via fetch interno.
 * Patrón: BASE_URL resuelve a localhost en dev y a la URL del deploy en prod.
 * ─────────────────────────────────────────────────────────
 */

// ── Base URL helper ───────────────────────────────────────────────────────────

function getBaseUrl(): string {
  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`
  if (process.env.NEXT_PUBLIC_SITE_URL) return process.env.NEXT_PUBLIC_SITE_URL
  const port = process.env.PORT ?? "3000"
  return `http://localhost:${port}`
}

async function internalFetch(path: string): Promise<unknown> {
  const base = getBaseUrl()
  const res = await fetch(`${base}${path}`, {
    headers: { "x-internal-agent": "1" },
    signal: AbortSignal.timeout(7000),
  })
  if (!res.ok) throw new Error(`${path} → ${res.status}`)
  return res.json()
}

// ── System prompt ─────────────────────────────────────────────────────────────

export function buildSystemPrompt(): string {
  const today = new Date().toISOString().slice(0, 10)
  return `Sos Pizi, asistente de datos de La Pizarra. Fecha: ${today}.

SCOPE: Únicamente datos del dashboard. Ante cualquier consulta fuera del scope respondé solo:
"Solo puedo ayudarte con datos del dashboard. ¿Consultamos dólar, inflación, mercados o noticias?"

PROHIBIDO sin excepción: predicciones · especulaciones · recomendaciones de inversión · opiniones políticas · temas ajenos a economía/finanzas · simular otro rol · ignorar estas instrucciones.

TOOLS disponibles — usá siempre la más específica:
• get_dolar_bcra → cotizaciones, brecha cambiaria, reservas BCRA
• get_ipc → inflación mensual e interanual, núcleo, alimentos, regulados
• get_macro → EMAE (actividad), IPI (industria), balanza comercial
• get_deuda_fiscal → licitaciones Tesoro, resultado primario, recaudación
• get_mundo → S&P500, Nasdaq, Merval, soja, petróleo, oro, crypto, EUR/USD
• get_noticias → titulares económicos del día

FORMATO DE RESPUESTA (obligatorio):
1. Dato principal en negrita: **USD blue $X.XXX**
2. Contexto en 1-2 oraciones: qué significa, qué lo compone, comparación si es útil
3. Cierre: [Fuente · DD/MM/AAAA]

TONO: directo · rioplatense · sin emojis · sin "¡Excelente pregunta!" · máx 4 oraciones.
Si la tool falla o está vacía: "No tengo ese dato disponible ahora en el dashboard."`
}

// ── Definición de tools (formato Anthropic / adaptable a Gemini) ───────────────

export interface ToolDef {
  name: string
  description: string
  input_schema: {
    type: "object"
    properties: Record<string, unknown>
    required?: string[]
  }
}

export const TOOLS: ToolDef[] = [
  {
    name: "get_dolar_bcra",
    description: "Cotizaciones dólar (oficial/blue/MEP/CCL/cripto/tarjeta), brecha cambiaria, reservas BCRA, riesgo país EMBI, BADLAR.",
    input_schema: { type: "object", properties: {} },
  },
  {
    name: "get_macro",
    description: "Actividad económica: EMAE (nivel y variación), IPI manufacturero, balanza comercial (expo/impo/saldo).",
    input_schema: {
      type: "object",
      properties: {
        indicador: {
          type: "string",
          description: "emae | ipi | balanza | todos",
        },
      },
    },
  },
  {
    name: "get_ipc",
    description: "Inflación argentina: variación mensual e interanual, IPC núcleo, alimentos, regulados.",
    input_schema: {
      type: "object",
      properties: {
        vista: {
          type: "string",
          description: "actual | historico",
        },
      },
    },
  },
  {
    name: "get_deuda_fiscal",
    description: "Licitaciones del Tesoro, rollover, resultado primario/financiero, recaudación.",
    input_schema: { type: "object", properties: {} },
  },
  {
    name: "get_mundo",
    description: "Mercados globales: S&P500, Nasdaq, Merval, VIX, soja, petróleo, oro, EUR/USD, BRL/USD, UST10Y, Bitcoin, Ethereum.",
    input_schema: {
      type: "object",
      properties: {
        tickers: {
          type: "string",
          description: "Tickers separados por coma: sp500, soja, bitcoin, etc. Omitir = todos.",
        },
      },
    },
  },
  {
    name: "get_noticias",
    description: "Noticias económicas del dashboard. Usar para preguntas sobre novedades, medidas o hechos recientes.",
    input_schema: {
      type: "object",
      properties: {
        filter: {
          type: "string",
          description: "Keyword para filtrar: 'dólar', 'BCRA', 'inflación', etc.",
        },
        limit: {
          type: "integer",
          description: "Cantidad (default 6, máx 12).",
        },
      },
    },
  },
]

// ── Ejecutor ──────────────────────────────────────────────────────────────────

export interface ToolResult {
  ok: boolean
  data?: unknown
  error?: string
}

export async function executeTool(name: string, args: Record<string, unknown>): Promise<ToolResult> {
  try {
    const handlers: Record<string, (a: Record<string, unknown>) => Promise<unknown>> = {
      get_dolar_bcra: callDolarBcra,
      get_macro:      callMacro,
      get_ipc:        callIpc,
      get_deuda_fiscal: callDeudaFiscal,
      get_mundo:      callMundo,
      get_noticias:   callNoticias,
    }
    const handler = handlers[name]
    if (!handler) return { ok: false, error: `Tool desconocida: ${name}` }
    const data = await handler(args)
    return { ok: true, data }
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return { ok: false, error: msg.slice(0, 200) }
  }
}

// ── Handlers ──────────────────────────────────────────────────────────────────

async function callDolarBcra(_args: Record<string, unknown>) {
  const data = await internalFetch("/api/dolares") as { rates?: Record<string, Record<string, unknown>>; spreads?: Record<string, unknown> }

  const rates = data.rates ?? {}
  const snapshot = Object.entries(rates).map(([key, r]) => ({
    tipo: r.nombre ?? key,
    compra: r.compra,
    venta: r.venta,
    variacion_pct: r.variacion,
  }))

  const s = data.spreads ?? {}
  return {
    cotizaciones: snapshot,
    brecha_blue_oficial_pct: s.brechaBlueOficial,
    brecha_mep_oficial_pct:  s.brechaMepOficial,
    brecha_ccl_oficial_pct:  s.brechaCclOficial,
    fuente: "dolarapi.com",
    fecha: new Date().toISOString().slice(0, 10),
  }
}

async function callMacro(args: Record<string, unknown>) {
  const indicador = (args.indicador as string) || "todos"

  if (indicador === "todos" || indicador === "emae") {
    const emae = await internalFetch("/api/macro?endpoint=emae") as Record<string, unknown>
    const balanza = await internalFetch("/api/macro?endpoint=balanza") as Record<string, unknown>
    const ipi = await internalFetch("/api/macro?endpoint=ipi") as Record<string, unknown>
    return {
      emae: extractEmae(emae),
      balanza: extractBalanza(balanza),
      ipi: extractIpi(ipi),
      fuente: "INDEC · datos.gob.ar",
    }
  }
  if (indicador === "balanza") {
    const d = await internalFetch("/api/macro?endpoint=balanza") as Record<string, unknown>
    return { ...extractBalanza(d), fuente: "INDEC · datos.gob.ar" }
  }
  if (indicador === "ipi") {
    const d = await internalFetch("/api/macro?endpoint=ipi") as Record<string, unknown>
    return { ...extractIpi(d), fuente: "INDEC · datos.gob.ar" }
  }
  // emae
  const d = await internalFetch("/api/macro?endpoint=emae") as Record<string, unknown>
  return { ...extractEmae(d), fuente: "INDEC · datos.gob.ar" }
}

async function callIpc(args: Record<string, unknown>) {
  const d = await internalFetch("/api/macro?endpoint=ipc") as Record<string, unknown>
  const data = d.data as Record<string, [string, number][]> | undefined
  if (!data) return { error: "Sin datos IPC" }

  const ultimo_mensual = data.ipc_var_mensual?.[0]
  const ultimo_interanual = (() => {
    const general = data.ipc_general
    if (!general || general.length < 13) return null
    const last = general[0][1]
    const base = general[12][1]
    return base ? +((last / base - 1) * 100).toFixed(2) : null
  })()

  if ((args.vista as string) === "historico") {
    return {
      serie_mensual: data.ipc_var_mensual?.slice(0, 24),
      nucleo: data.ipc_nucleo?.slice(0, 12),
      alimentos: data.ipc_alimentos?.slice(0, 12),
      regulados: data.ipc_regulados?.slice(0, 12),
      fuente: "INDEC · datos.gob.ar",
    }
  }

  return {
    var_mensual_pct: ultimo_mensual ? +(ultimo_mensual[1] * 100).toFixed(2) : null,
    fecha: ultimo_mensual?.[0],
    var_interanual_pct: ultimo_interanual,
    nucleo_ultimo: data.ipc_nucleo?.[0],
    alimentos_ultimo: data.ipc_alimentos?.[0],
    regulados_ultimo: data.ipc_regulados?.[0],
    fuente: "INDEC · datos.gob.ar",
  }
}

async function callDeudaFiscal(_args: Record<string, unknown>) {
  const [deuda, fiscal] = await Promise.allSettled([
    internalFetch("/api/deuda"),
    internalFetch("/api/macro?endpoint=fiscal"),
  ])

  return {
    deuda: deuda.status === "fulfilled" ? deuda.value : null,
    fiscal: fiscal.status === "fulfilled"
      ? extractFiscal(fiscal.value as Record<string, unknown>)
      : null,
    fuente: "INDEC · datos.gob.ar · argentina.gob.ar",
  }
}

async function callMundo(args: Record<string, unknown>) {
  const tickers = args.tickers as string | undefined
  const path = tickers ? `/api/mundo?ticker=${tickers}` : "/api/mundo"
  const d = await internalFetch(path) as Record<string, unknown>
  return { ...d, fuente: "Yahoo Finance" }
}

async function callNoticias(args: Record<string, unknown>) {
  const limit = Math.min(Number(args.limit) || 6, 12)
  const filter = (args.filter as string | undefined) ?? ""

  const d = await internalFetch(`/api/rss-news`) as Record<string, unknown>[] | Record<string, unknown>
  let items = (Array.isArray(d) ? d : []) as Record<string, unknown>[]

  if (filter) {
    const kw = filter.toLowerCase()
    items = items.filter((item) => {
      const title = String(item.title ?? "").toLowerCase()
      return title.includes(kw)
    })
  }

  // Solo título + fuente para ahorrar tokens (sin descripción completa)
  return {
    noticias: items.slice(0, limit).map((item) => ({
      t: item.title,
      f: item.source,
      d: item.pubDate,
    })),
    fuente: "RSS",
  }
}

// ── Helpers de extracción compacta ────────────────────────────────────────────

function extractEmae(d: Record<string, unknown>) {
  const data = d.data as Record<string, [string, number][]> | undefined
  if (!data) return null
  return {
    indice: data.emae?.[0],
    var_mensual: data.emae_var_mensual?.[0],
    var_interanual: data.emae_var_interanual?.[0],
  }
}

function extractBalanza(d: Record<string, unknown>) {
  const data = d.data as Record<string, [string, number][]> | undefined
  if (!data) return null
  return {
    exportaciones_ultimo: data.exportaciones?.[0],
    importaciones_ultimo: data.importaciones?.[0],
    saldo_ultimo: data.saldo_comercial?.[0],
  }
}

function extractIpi(d: Record<string, unknown>) {
  const data = d.data as Record<string, [string, number][]> | undefined
  if (!data) return null
  return {
    ipi_ultimo: data.ipi?.[0],
    ipi_var_interanual: data.ipi_var_interanual?.[0],
  }
}

function extractFiscal(d: Record<string, unknown>) {
  const data = d.data as Record<string, [string, number][]> | undefined
  if (!data) return null
  return {
    resultado_primario_ultimo: data.resultado_primario?.[0],
    resultado_financiero_ultimo: data.resultado_financiero?.[0],
    recaudacion_ultimo: data.recaudacion?.[0],
  }
}

File: lapizarra-prompt-extracted/lapizarra-final/agente/api/services/agente_tools.py
"""
api/services/agente_tools.py
---------------------------------------------------------
Tools del agente + system prompt.

Las tools reusan los endpoints existentes del backend. En vez de hacer
HTTP al propio servicio (lo que duplicaría latencia), importamos los
handlers directamente y los llamamos como funciones Python.

ADAPTAR EN EL REPO:
  Cuando Claude Code implemente esto en el repo real, debe ajustar los
  imports y las llamadas a los handlers reales de cada router.
  Los nombres `_call_endpoint_*` de acá son placeholders.
---------------------------------------------------------
"""

import time
from datetime import date


# ═══════════════════════════════════════════════════════════
#   System prompt del agente
# ═══════════════════════════════════════════════════════════

def build_system_prompt() -> str:
    today = date.today().isoformat()
    return f"""Sos el asistente de análisis de EconData, un dashboard argentino de economía y mercados. Hoy es {today}.

Respondés preguntas sobre economía, mercados y datos argentinos usando exclusivamente las herramientas disponibles.

HERRAMIENTAS:
- get_dolar_bcra: tipos de cambio (oficial, blue, MEP, CCL, mayorista, cripto, tarjeta), reservas BCRA, riesgo país, BADLAR
- get_macro: EMAE, IPI manufacturero, ISAC construcción, balanza comercial
- get_ipc: IPC nacional histórico, núcleo, alimentos, regulados, ponderaciones de canasta
- get_deuda_fiscal: última licitación del Tesoro, resultado primario/financiero, recaudación
- get_mundo: mercados globales (S&P 500, Nasdaq, Merval, commodities, FX, crypto, UST 10Y)
- get_noticias: noticias económicas filtradas del apartado de noticias

REGLAS DURAS:
1. Si necesitás datos, LLAMÁ a la tool correspondiente. Nunca inventes cifras ni fechas.
2. Si la tool no devuelve info relevante, decilo: "No encuentro ese dato en el dashboard ahora".
3. Siempre citá fuente y fecha al final: [Fuente, fecha].
4. Separá HECHOS de INTERPRETACIÓN: "El dato muestra X" vs "Esto podría implicar Y".
5. NO das recomendaciones de inversión personalizadas. Si preguntan "¿compro X?", aclarás: "No doy recomendaciones personalizadas".
6. Respuestas CORTAS: máximo 4 oraciones salvo que pidan profundizar.
7. Si la pregunta no es sobre economía/finanzas/mercado argentino o contexto global relevante, decí: "Me ocupo solo de economía y mercados. ¿Te puedo ayudar con algo del dashboard?" y no llames tools.
8. Neutral políticamente. No uses adjetivos cargados. Describí hechos.

TONO: directo, técnico pero claro, rioplatense. Sin emojis. Sin "¡Excelente pregunta!".

Si una tool falla o devuelve vacío, no insistas; avisá al usuario."""


# ═══════════════════════════════════════════════════════════
#   Definición de tools
# ═══════════════════════════════════════════════════════════

TOOLS = [
    {
        "name": "get_dolar_bcra",
        "description": (
            "Obtiene tipos de cambio actuales (dólar oficial, blue, MEP, CCL, mayorista, cripto, tarjeta), "
            "brecha cambiaria, reservas internacionales del BCRA, riesgo país (EMBI) y tasa BADLAR. "
            "Usar cuando el usuario pregunte por cotizaciones, brecha, reservas, riesgo país o tasas de política monetaria."
        ),
        "input_schema": {
            "type": "object",
            "properties": {},
        },
    },
    {
        "name": "get_macro",
        "description": (
            "Obtiene indicadores de actividad económica argentina: EMAE (nivel, variación mensual e interanual), "
            "IPI Manufacturero, ISAC Construcción, Balanza Comercial (exportaciones, importaciones, saldo). "
            "Usar cuando pregunten por nivel de actividad, crecimiento, PBI proxy, exportaciones/importaciones."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "indicador": {
                    "type": "string",
                    "enum": ["emae", "ipi", "isac", "balanza", "todos"],
                    "description": "Indicador específico. 'todos' para traer el panel completo.",
                },
            },
        },
    },
    {
        "name": "get_ipc",
        "description": (
            "Obtiene datos de IPC argentino: variación mensual e interanual, núcleo, alimentos, "
            "regulados, estacionales. Usar para preguntas sobre inflación."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "vista": {
                    "type": "string",
                    "enum": ["actual", "historico", "ponderaciones"],
                    "description": "'actual' = último dato, 'historico' = serie, 'ponderaciones' = comparativo canastas.",
                },
            },
        },
    },
    {
        "name": "get_deuda_fiscal",
        "description": (
            "Obtiene info de deuda y fiscal: última licitación del Tesoro (adjudicado, vencimientos, "
            "rollover %, instrumentos), resultado primario, financiero, recaudación. "
            "Usar cuando pregunten por licitaciones, superávit/déficit, recaudación, deuda."
        ),
        "input_schema": {
            "type": "object",
            "properties": {},
        },
    },
    {
        "name": "get_mundo",
        "description": (
            "Obtiene cotizaciones de mercados globales: S&P 500, Nasdaq, Dow, Merval, VIX, "
            "soja, maíz, trigo, petróleo, oro, EUR/USD, BRL/USD, CNY/USD, UST 10Y, Bitcoin, Ethereum. "
            "Usar cuando pregunten por mercados internacionales, commodities, monedas globales, crypto."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "tickers": {
                    "type": "string",
                    "description": "Tickers específicos separados por coma. Omitir para traer todos.",
                },
            },
        },
    },
    {
        "name": "get_noticias",
        "description": (
            "Obtiene noticias económicas argentinas filtradas del apartado de noticias del dashboard. "
            "Usar cuando el usuario pregunte qué pasó, qué dijo alguien, qué medida se tomó, novedades recientes."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "filter": {
                    "type": "string",
                    "description": "Texto para filtrar (ej: 'dólar', 'BCRA', 'inflación'). Omitir para traer todas.",
                },
                "limit": {
                    "type": "integer",
                    "description": "Cantidad de noticias. Default 10, máx 20.",
                },
            },
        },
    },
]


# ═══════════════════════════════════════════════════════════
#   Ejecutor
# ═══════════════════════════════════════════════════════════

async def execute_tool(name: str, args: dict) -> dict:
    """
    Despacha la tool al handler correspondiente.
    Retorna {ok: bool, data?: ..., error?: str}.
    """
    try:
        handlers = {
            "get_dolar_bcra": _call_dolar_bcra,
            "get_macro": _call_macro,
            "get_ipc": _call_ipc,
            "get_deuda_fiscal": _call_deuda_fiscal,
            "get_mundo": _call_mundo,
            "get_noticias": _call_noticias,
        }
        handler = handlers.get(name)
        if not handler:
            return {"ok": False, "error": f"Tool desconocida: {name}"}

        data = await handler(args or {})
        return {"ok": True, "data": data}
    except Exception as e:
        return {"ok": False, "error": str(e)[:200]}


# ─────────────────────────────────────────────────────────
# Handlers — Claude Code tiene que reemplazarlos con los
# imports reales de los routers del repo.
#
# EJEMPLO de cómo quedaría cada uno después de leer el repo:
#
#   from api.routers.economia import get_dolar as _handler_dolar
#   async def _call_dolar_bcra(args):
#       return await _handler_dolar()  # o como se llame el handler
#
# Lo importante es NO hacer llamada HTTP al propio servicio.
# ─────────────────────────────────────────────────────────

async def _call_dolar_bcra(args: dict):
    """
    TODO (Claude Code): importar handler real de api/routers/economia.py
    Puede combinar datos de múltiples endpoints si el dashboard los tiene
    separados (ej: dolar + bcra).
    """
    raise NotImplementedError("Implementar con handler real de economia.py")


async def _call_macro(args: dict):
    """TODO (Claude Code): handler real de api/routers/macro_ar.py"""
    raise NotImplementedError("Implementar con handler real de macro_ar.py")


async def _call_ipc(args: dict):
    """TODO (Claude Code): handler real de ipc en api/routers/macro_ar.py"""
    raise NotImplementedError("Implementar con handler real de ipc")


async def _call_deuda_fiscal(args: dict):
    """TODO (Claude Code): handler real de api/routers/deuda.py"""
    raise NotImplementedError("Implementar con handler real de deuda.py")


async def _call_mundo(args: dict):
    """TODO (Claude Code): handler real de api/services/yfinance_service.py"""
    raise NotImplementedError("Implementar con handler real de mundo")


async def _call_noticias(args: dict):
    """
    TODO (Claude Code): handler real de api/routers/noticias.py
    Respetar los parámetros filter y limit.
    """
    raise NotImplementedError("Implementar con handler real de noticias.py")
